### PR TITLE
feature: add distribution policy for release-1.21

### PR DESCRIPTION
**Pull request by qiuming520**
#### What type of PR is this?

/kind feature
#### What does this PR do?

scheduler plugin v1.21.5 -- Combined with distribution policy to finely distribute user resources to leaf nodes or host nodes
#### Which issue(s) does this PR fix?

Fixes  https://github.com/kosmos-io/kosmos/issues/297
#### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

File: deploy/scheduler/deployment.yaml [hunk 1]
```diff
@@ -98,3 +99,7 @@ data:
           - name: KnodeVolumeBinding
             args:
               bindTimeoutSeconds: 5
+          - name: KNodeDistribution
+            args:
+            # kubeConfigPath: "REPLACE_ME_WITH_KUBE_CONFIG_PATH"
+              kubeConfigPath: "/etc/kubernetes/scheduler.conf"
```
Comment (duanmengkk) [inline deploy/scheduler/deployment.yaml]: add a new line

File: hack/update-crds.sh [hunk 2]
```diff
@@ -14,4 +14,4 @@ export PATH=$PATH:$GOPATH/bin
 
 controller-gen crd paths=./pkg/apis/kosmos/... output:crd:dir="${REPO_ROOT}/deploy/crds"
 
-go run "${REPO_ROOT}/hack/generate/generate.go"
\ No newline at end of file
+#go run "${REPO_ROOT}/hack/generate/generate.go"
```
Comment (duanmengkk) [inline hack/update-crds.sh]: add a new line and why comment the line

Comment (qiuming520) [reply hack/update-crds.sh]: have done

File: deploy/scheduler/deployment.yaml (hunk 1 above)
Comment (qiuming520) [reply deploy/scheduler/deployment.yaml]: have done

File: deploy/scheduler/deployment.yaml [hunk 3]
```diff
@@ -98,3 +99,7 @@ data:
           - name: KnodeVolumeBinding
             args:
               bindTimeoutSeconds: 5
+        #  - name: KNodeDistribution
+        #    args:
+        #    # kubeConfigPath: "REPLACE_ME_WITH_KUBE_CONFIG_PATH"
+        #      kubeConfigPath: "/etc/kubernetes/scheduler.conf"
```
Comment (yuleichun-striving) [inline deploy/scheduler/deployment.yaml]: The code here is in comment status, is it normal?

File: pkg/apis/config/register.go [hunk 4]
```diff
@@ -38,6 +38,7 @@ var (
 func addKnownTypes(scheme *runtime.Scheme) error {
 	scheme.AddKnownTypes(SchemeGroupVersion,
 		&KnodeVolumeBindingArgs{},
```
Comment (yuleichun-striving) [inline pkg/apis/config/register.go]: Be consistent with capitalization of "Knode" code

File: pkg/apis/config/register.go [hunk 5]
```diff
@@ -47,4 +48,5 @@ func init() {
 	// generated functions takes place in the generated files. The separation
 	// makes the code compile even when the generated files are missing.
 	localSchemeBuilder.Register(addKnownTypes)
```
Comment (yuleichun-striving) [inline pkg/apis/config/register.go:50]: Keep the initial case of words in the code consistent

File: pkg/apis/kosmos/v1alpha1/distributionpolicy_types.go [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+	LEAFNODE NodeType = "leaf"
+
+	//MIXNODE represents the host nodes and child nodes
+	MIXNODE NodeType = "mix"
+)
+
+type PolicyTerm struct {
+	// +required
+	Name string `json:"name,omitempty"`
+
```
Comment (duanmengkk) [inline pkg/apis/kosmos/v1alpha1/distributionpolicy_types.go:86]: follow this
  https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#optional-vs-required

File: pkg/apis/kosmos/v1alpha1/distributionpolicy_types.go [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+	// Filter resource by labelSelector
+	// If target resource name is not empty, labelSelector will be ignored.
+	// +optional
+	LabelSelector *metav1.LabelSelector `json:"labelSelector,omitempty"`
+
+	// Name of the Policy.
+	// +required
+	PolicyName string `json:"policyName,omitempty"`
```
Comment (duanmengkk) [inline pkg/apis/kosmos/v1alpha1/distributionpolicy_types.go]: Is NameSpace Prefix needed ?   @hanweisen

File: pkg/apis/config/register.go (hunk 5 above)
Comment (qiuming520) [reply pkg/apis/config/register.go:50]: This is automatically generated code. The case of method names controls visibility.

File: deploy/scheduler/deployment.yaml (hunk 3 above)
Comment (qiuming520) [reply deploy/scheduler/deployment.yaml]: yes~,You can refer to [scheduler-plugins](https://github.com/kubernetes-sigs/scheduler-plugins/blob/release-1.21/manifests/noderesourcetopology/scheduler-config.yaml#L6)

File: pkg/apis/kosmos/v1alpha1/distributionpolicy_types.go (hunk 7 above)
Comment (qiuming520) [reply pkg/apis/kosmos/v1alpha1/distributionpolicy_types.go]: > Is NameSpace Prefix needed ? @hanweisen

yes~we will add labelselector and namespace  prefix  terms

File: pkg/apis/kosmos/v1alpha1/distributionpolicy_types.go (hunk 6 above)
Comment (qiuming520) [reply pkg/apis/kosmos/v1alpha1/distributionpolicy_types.go:86]: > follow this https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#optional-vs-required

yes,have done

**Pull request merged**